### DISCUSSION TITLE
[mini.visits] Ability to delete multiple registered paths via picker

### DISCUSSION THREAD

**Discussion by karlskewes** [nvim-mini/mini.nvim#1003] [Ideas]
First, thank you for writing such a great plugin. It has been smooth sailing integrating and using it every day.

There's one part of my workflow that I haven't figured out with the default features/examples and I'm open to either making something or changing my workflow.
The feature is deleting multiple registered paths using a picker of some sort.

### My workflow (harpoon-bred) 

1. disable automatic path registration - `track = { event = '' }`
2. browse/jump to file
3. add visit - `MiniVisits.register_visit()`
4. iterate registered paths - `MiniVisits.iterate_paths("forward", nil, { wrap = true })` (forward/backward)
5. repeat 2-4
6. finished with files (may have been across multiple git branches)
7. delete (some|all) registered visits - `MiniVisits.remove_path()`

### Hiccup

In Step 7 above, I need to visit each file I want to remove/unregister and then remove it.
It would be great if I could delete multiple in one go.

### Harpoon feature

A custom picker can be opened which supports vim actions `dd, p, etc` to edit/reorder saved paths. 
I would often `dd` in the single picker multiple times when finished with multiple files.

### vim.ui.select

I see it's possible to customize the action that `vim.ui.select()` does and have used it below.
This enables deleting a single path via a picker, but I'm not sure how to delete multiple here.

```lua
vim.keymap.set(
  'n', 
  '<leader>vd', 
  function()
    vim.ui.select(MiniVisits.list_paths(), {}, function(choice)
      MiniVisits.remove_path(choice)
    end)
  end, 
  { desc = '[V]isits [d]elete paths' },
)
```

### telescope.nvim

This seems to support multiple selections with the [default-mappings](https://github.com/nvim-telescope/telescope.nvim?tab=readme-ov-file#default-mappings) `Tab`.

I tried using https://github.com/nvim-telescope/telescope-ui-select.nvim as the picker and can use `Tab` to "mark" multiple paths.

However I can't see how multiple "marked paths" can be passed to the function, it seems to pass only what I hit `<CR>` on, which makes sense as the function definition is `vim.ui.select(fn, {}, <single_choice>)`.

I've come across a potential telescope based solution which I could change the action from opening file to deleting path: https://github.com/nvim-telescope/telescope.nvim/issues/1048#issuecomment-1679797700

### mini.pick

Is there something in https://github.com/echasnovski/mini.pick that would work?

> Mark/unmark items to choose later.

How would I apply a command to all "marked" items?

**Comment by echasnovski** [comment]
Thanks for the idea and kind words about 'mini.visits'!

Explicit index manipulation can be done in at least the following ways:
- Open index file *outside of Neovim* (or at least with disabled 'mini.visits' because it will autowrite before Neovim is closed) and edit it directly. This is the most straightforward approach I'd recommend.
- Programmatically it can be done by getting whole index with [`get_index()`](https://github.com/echasnovski/mini.nvim/blob/35422e90b0de232245e2b1d3b3737a7c419d377b/doc/mini-visits.txt#L716), manipulate it somehow, and set it back with [`set_index()`](https://github.com/echasnovski/mini.nvim/blob/35422e90b0de232245e2b1d3b3737a7c419d377b/doc/mini-visits.txt#L724).

------

Indeed, applying a command to more than one item is not a part of `vim.ui.select()` interface. But it can be for its implementation, which is indeed a one from 'mini.pick'. There are at least these ways to perform action on all marked items:
- Define picker's [`choose_marked()`](https://github.com/echasnovski/mini.nvim/blob/35422e90b0de232245e2b1d3b3737a7c419d377b/doc/mini-pick.txt#L426) method which will be called after pressing `<CR>` with array of marked items as argument.
- Define [custom action](https://github.com/echasnovski/mini.nvim/blob/35422e90b0de232245e2b1d3b3737a7c419d377b/doc/mini-pick.txt#L586) to be invoked on certain key (like, for example, `<C-d>`). Inside its `func` method it can call [`get_picker_matches()`](https://github.com/echasnovski/mini.nvim/blob/35422e90b0de232245e2b1d3b3737a7c419d377b/doc/mini-pick.txt#L1225) and use `marked` field of its output to get marked items and do something based on that information.

------

Hope this helps.

**Comment by karlskewes** [comment]
Thanks very much Evgeni!

I ended up with the following which enables deleting current and marked items using the default `mini.pick` key mappings.
The list is refreshed after each action, `Esc` to exit picker.
```lua
      vim.keymap.set(
        'n',
        '<leader>vd',
        function()
          local picker = require('mini.pick')
          local visits = require('mini.visits')

          local remove_paths = function(items)
            for _, v in ipairs(items) do
              visits.remove_path(v)
            end

            picker.set_picker_items(visits.list_paths())
            return true
          end

          picker.setup({
            source = {
              items = visits.list_paths(),
              choose = function(item) return remove_paths({ item }) end,
              choose_marked = remove_paths,
            }
          })

          picker.start()
        end,
        { desc = '[V]isits [d]elete paths' }
      )
```

**Reply by echasnovski** [reply]
Nice! Something like this is indeed a preferred approach to handling this use case.

**Comment by krovuxdev** [comment]
Interesting! I also found an alternative approach that I thought might be worth sharing. I mentioned it [here](https://github.com/echasnovski/mini.nvim/discussions/1752#discussioncomment-12888178), but I'm reposting it for better visibility.

You can override the `visit_paths` picker like this to allow removing a buffer from the list with `<C-d>`:

```lua
visit_paths = function(local_opts, opts)
  local_opts = local_opts or {}
  local wipeout_cur = function()
    MiniVisits.remove_path(MiniPick.get_picker_matches().current)
    MiniExtra.pickers.visit_paths(local_opts, opts)
  end
  local visit_mappings = {
    wipeout = {
      char = "<C-d>",
      func = wipeout_cur,
    },
  }
  opts = vim.tbl_deep_extend("force", { mappings = visit_mappings }, opts or {})
  return MiniExtra.pickers.visit_paths(local_opts, opts)
end,
```

With this setup, you can press `<C-d>` to remove the current buffer from the picker and it will automatically refresh. I found it to be a simple and convenient solution.

**Reply by echasnovski** [reply]
> ```lua
> -- ...
>     MiniVisits.remove_path(MiniPick.get_picker_matches().current.bufnr)
> -- ...
> ```

I don't think `.bufnr` is correct here. It should probably be just `MiniPick.get_picker_matches().current`.

**Reply by krovuxdev** [reply]
What's the difference between `bufnr` and `current`? It works for me with just `bufnr` 🫠 I'll update it in a bit.

**Reply by echasnovski** [reply]
> What's the difference between `bufnr` and `current`? It works for me with just `bufnr` 🫠 I'll update it in a bit.

Items in `MiniExtra.pickers.visit_paths` should be strings with absolute paths. This is what `MiniVisits.remove_path()` expects as its first argument. Having `.bufnr` means that items are tables.

**Reply by krovuxdev** [reply]
Thanks a lot, I’ve updated it now!

**Comment by abeldekat** [comment]
Your workflow resembles mine. When `harpoon2` development halted, I configured `mini.visits` to operate like `harpoon2`. Which resulted in an [internal](https://github.com/abeldekat/nvim_pde/blob/603c86b11188a9605e7acbdfeb9b52429c8bc868/lua/ak/mini/visits_harpooned.lua#L13) plugin-like structure I have been [using](https://github.com/abeldekat/nvim_pde/blob/603c86b11188a9605e7acbdfeb9b52429c8bc868/lua/ak/config/editor/visits.lua#L2) ever since. 

If you are interested, here is the corresponding [post](https://github.com/echasnovski/mini.nvim/discussions/1158) in discussions. 

I don't use a picker to delete paths. Instead, I took inspiration from `mini.deps`, opening a buffer in a new tab. I can than do `dG` and save.
